File: oprel/webui-react/services/mcp.ts
import { API_BASE } from './api';

const MCP_BASE = () => `${API_BASE}/mcp`;

export interface Connector {
  id: string;
  name: string;
  description: string;
  category: string;
  icon: string;
  color: string;
  auth_type: string;
  auth_label: string;
  auth_help: string;
  auth_url?: string;
  tools_preview: string[];
  requires_package?: string;
  config_fields?: Array<{
    name: string;
    label: string;
    type: string;
    required?: boolean;
    help?: string;
  }>;
}

export interface ConnectorInstance {
  id: string;
  builtin_id: string;
  name: string;
  transport: string;
  config: Record<string, string>;
  enabled: boolean;
  status: 'disconnected' | 'connecting' | 'connected' | 'error';
  error?: string;
}

export const mcpApi = {
  // Catalog
  getCatalog: async (): Promise<{ connectors: Connector[] }> => {
    const res = await fetch(`${MCP_BASE()}/catalog`);
    if (!res.ok) throw new Error('Failed to fetch connector catalog');
    return res.json();
  },

  // Connector CRUD
  listConnectors: async (): Promise<{ connectors: ConnectorInstance[] }> => {
    const res = await fetch(`${MCP_BASE()}/connectors`);
    if (!res.ok) throw new Error('Failed to list connectors');
    return res.json();
  },

  addConnector: async (builtin_id: string, config: Record<string, string>, name?: string, enabled = true): Promise<{ connector: ConnectorInstance }> => {
    const res = await fetch(`${MCP_BASE()}/connectors`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ builtin_id, config, name, enabled }),
    });
    if (!res.ok) throw new Error('Failed to add connector');
    return res.json();
  },

  updateConnector: async (id: string, patch: Partial<ConnectorInstance>): Promise<{ connector: ConnectorInstance }> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(patch),
    });
    if (!res.ok) throw new Error('Failed to update connector');
    return res.json();
  },

  deleteConnector: async (id: string): Promise<void> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    });
    if (!res.ok) throw new Error('Failed to delete connector');
  },

  // Lifecycle
  connectConnector: async (id: string): Promise<any> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}/connect`, {
      method: 'POST',
    });
    if (!res.ok) throw new Error('Failed to connect connector');
    return res.json();
  },

  disconnectConnector: async (id: string): Promise<any> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}/disconnect`, {
      method: 'POST',
    });
    if (!res.ok) throw new Error('Failed to disconnect connector');
    return res.json();
  },

  reconnectConnector: async (id: string): Promise<any> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}/reconnect`, {
      method: 'POST',
    });
    if (!res.ok) throw new Error('Failed to reconnect connector');
    return res.json();
  },

  testConnector: async (id: string): Promise<any> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}/test`, {
      method: 'POST',
    });
    if (!res.ok) throw new Error('Failed to test connector');
    return res.json();
  },

  // Tools & status
  getConnectorTools: async (id: string): Promise<{ tools: any[]; count: number }> => {
    const res = await fetch(`${MCP_BASE()}/connectors/${encodeURIComponent(id)}/tools`);
    if (!res.ok) throw new Error('Failed to fetch connector tools');
    return res.json();
  },

  getStatus: async (): Promise<{ total_connectors: number; connected: number; disconnected: number; total_tools: number; live_connector_ids: string[] }> => {
    const res = await fetch(`${MCP_BASE()}/status`);
    if (!res.ok) throw new Error('Failed to fetch MCP status');
    return res.json();
  },
};

File: oprel/server/routes/mcp.py
from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, BackgroundTasks, HTTPException, Request
from fastapi.responses import HTMLResponse
from pydantic import BaseModel

from oprel.server import db
from oprel.mcp import manager as mcp_manager
from oprel.mcp import oauth as mcp_oauth
from oprel.mcp.registry import get_builtin_connector, list_builtin_connectors

router = APIRouter(prefix="/mcp", tags=["mcp"])


# ── request schemas ───────────────────────────────────────────────────────────

class CreateConnectorRequest(BaseModel):
    builtin_id: str
    name: str | None = None
    config: dict[str, Any] = {}
    enabled: bool = True


class UpdateConnectorRequest(BaseModel):
    name: str | None = None
    config: dict[str, Any] | None = None
    enabled: bool | None = None


# ── catalog ───────────────────────────────────────────────────────────────────

@router.get("/catalog")
async def get_catalog():
    return {"connectors": list_builtin_connectors()}


@router.get("/catalog/{builtin_id}")
async def get_catalog_item(builtin_id: str):
    builtin = get_builtin_connector(builtin_id)
    if not builtin:
        raise HTTPException(404, f"No built-in connector '{builtin_id}'")
    return builtin


# ── connector instances ───────────────────────────────────────────────────────

@router.get("/connectors")
async def list_connectors():
    instances = db.list_mcp_connectors()
    live = set(mcp_manager.get_all_live_connector_ids())
    for inst in instances:
        if inst["id"] in live and inst.get("status") != "connected":
            inst["status"] = "connected"
    return {"connectors": instances}


@router.post("/connectors", status_code=201)
async def add_connector(body: CreateConnectorRequest, background_tasks: BackgroundTasks):
    builtin = get_builtin_connector(body.builtin_id)
    if not builtin and body.builtin_id != "__custom__":
        raise HTTPException(400, f"Unknown built-in connector id '{body.builtin_id}'")

    connector_id = body.builtin_id if body.builtin_id != "__custom__" else f"custom_{uuid.uuid4().hex[:8]}"
    if db.get_mcp_connector(connector_id):
        connector_id = f"{connector_id}_{uuid.uuid4().hex[:6]}"

    template = builtin or {}
    data = {
        "id": connector_id,
        "builtin_id": body.builtin_id,
        "name": body.name or template.get("name", connector_id),
        "transport": template.get("transport", body.config.get("transport", "stdio")),
        "config": body.config,
        "enabled": body.enabled,
        "status": "disconnected",
    }
    connector = db.upsert_mcp_connector(data)

    if body.enabled:
        background_tasks.add_task(mcp_manager.connect, connector_id)

    return {"connector": connector}


@router.get("/connectors/{connector_id}")
async def get_connector(connector_id: str):
    connector = db.get_mcp_connector(connector_id)
    if not connector:
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    return {"connector": connector}


@router.patch("/connectors/{connector_id}")
async def update_connector(
    connector_id: str, body: UpdateConnectorRequest, background_tasks: BackgroundTasks
):
    existing = db.get_mcp_connector(connector_id)
    if not existing:
        raise HTTPException(404, f"Connector '{connector_id}' not found")

    updated = dict(existing)
    if body.name is not None:
        updated["name"] = body.name
    if body.config is not None:
        updated["config"] = {**existing.get("config", {}), **body.config}

    was_enabled = existing.get("enabled", True)
    if body.enabled is not None:
        updated["enabled"] = body.enabled

    connector = db.upsert_mcp_connector(updated)

    if body.enabled is True and not was_enabled:
        background_tasks.add_task(mcp_manager.connect, connector_id)
    elif body.enabled is False and was_enabled:
        background_tasks.add_task(mcp_manager.disconnect, connector_id)
    elif body.config is not None:
        background_tasks.add_task(mcp_manager.reconnect, connector_id)

    return {"connector": connector}


@router.delete("/connectors/{connector_id}", status_code=204)
async def delete_connector(connector_id: str, background_tasks: BackgroundTasks):
    if not db.get_mcp_connector(connector_id):
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    background_tasks.add_task(mcp_manager.disconnect, connector_id)
    db.delete_mcp_connector(connector_id)
    return None


# ── connection control ────────────────────────────────────────────────────────

@router.post("/connectors/{connector_id}/connect")
async def connect_connector(connector_id: str, background_tasks: BackgroundTasks):
    if not db.get_mcp_connector(connector_id):
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    background_tasks.add_task(mcp_manager.connect, connector_id)
    return {"status": "connecting", "connector_id": connector_id}


@router.post("/connectors/{connector_id}/disconnect")
async def disconnect_connector(connector_id: str, background_tasks: BackgroundTasks):
    background_tasks.add_task(mcp_manager.disconnect, connector_id)
    return {"status": "disconnecting", "connector_id": connector_id}


@router.post("/connectors/{connector_id}/reconnect")
async def reconnect_connector(connector_id: str, background_tasks: BackgroundTasks):
    if not db.get_mcp_connector(connector_id):
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    background_tasks.add_task(mcp_manager.reconnect, connector_id)
    return {"status": "reconnecting", "connector_id": connector_id}


@router.post("/connectors/{connector_id}/test")
async def test_connector(connector_id: str):
    if not db.get_mcp_connector(connector_id):
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    return await mcp_manager.test_connection(connector_id)


# ── tools ─────────────────────────────────────────────────────────────────────

@router.get("/connectors/{connector_id}/tools")
async def get_connector_tools(connector_id: str):
    if not db.get_mcp_connector(connector_id):
        raise HTTPException(404, f"Connector '{connector_id}' not found")
    tools = db.get_mcp_tools(connector_id)
    return {"connector_id": connector_id, "tools": tools, "count": len(tools)}


@router.get("/tools")
async def get_all_tools():
    tools = db.get_all_mcp_enabled_tools()
    return {"tools": tools, "count": len(tools)}


# ── audit log & status ────────────────────────────────────────────────────────

@router.get("/logs")
async def get_call_logs(limit: int = 50, conversation_id: str | None = None):
    logs = db.list_mcp_call_logs(limit=limit, conversation_id=conversation_id)
    return {"logs": logs, "count": len(logs)}


@router.get("/status")
async def get_mcp_status():
    connectors = db.list_mcp_connectors()
    live_ids = set(mcp_manager.get_all_live_connector_ids())
    tools = db.get_all_mcp_enabled_tools()
    return {
        "total_connectors": len(connectors),
        "connected": sum(1 for c in connectors if c["id"] in live_ids),
        "disconnected": sum(1 for c in connectors if c["id"] not in live_ids),
        "total_tools": len(tools),
        "live_connector_ids": list(live_ids),
    }


# ── oauth ─────────────────────────────────────────────────────────────────────

@router.get("/connectors/{connector_id}/oauth/start")
async def oauth_start(connector_id: str, request: Request):
    """
    Returns the authorization URL. Frontend opens this in a popup/redirect.
    """
    connector = db.get_mcp_connector(connector_id)
    if not connector:
        raise HTTPException(404, "Connector not found")

    builtin = get_builtin_connector(connector["builtin_id"]) or {}
    oauth_cfg = builtin.get("oauth")
    if not oauth_cfg:
        raise HTTPException(400, "This connector does not use OAuth")

    redirect_uri = str(request.base_url) + "api/mcp/oauth/callback"
    auth_url, state = mcp_oauth.build_auth_url(connector_id, oauth_cfg, redirect_uri)

    return {"auth_url": auth_url, "state": state}


@router.get("/oauth/callback")
async def oauth_callback(
    code: str,
    state: str,
    background_tasks: BackgroundTasks,
):
    """
    Provider redirects here after user approves.
    Exchanges code for tokens, stores them, reconnects the session.
    """
    # Recover which connector this belongs to
    pending = mcp_oauth._pending.get(state)
    if not pending:
        raise HTTPException(400, "Invalid or expired OAuth state")

    connector_id = pending["connector_id"]
    connector    = db.get_mcp_connector(connector_id)
    builtin      = get_builtin_connector(connector["builtin_id"]) or {}
    oauth_cfg    = builtin.get("oauth", {})

    try:
        tokens = await mcp_oauth.exchange_code(state, code, oauth_cfg)
    except Exception as exc:
        raise HTTPException(400, f"Token exchange failed: {exc}")

    # Store tokens in connector config
    updated_cfg = {
        **connector.get("config", {}),
        "oauth_token":         tokens["access_token"],
        "oauth_refresh_token": tokens.get("refresh_token", ""),
        "oauth_expires_at":    tokens["_expires_at"],
    }
    db.upsert_mcp_connector({**connector, "config": updated_cfg})

    # Reconnect the live session with the new token
    background_tasks.add_task(mcp_manager.reconnect, connector_id)

    # Close the popup / redirect to settings page
    return HTMLResponse("""
        <script>
            window.opener?.postMessage({type:'oauth_success'}, '*');
            window.close();
        </script>
        <p>Connected! You can close this tab.</p>
    """)
